src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosContainerProvider.cs

```diff
@@ -51,14 +58,20 @@ public class CosmosContainerProvider : IHostedService, IRequireInitializationOnF
 
             string collectionId = collectionConfiguration.Get(Constants.CollectionConfigurationName).CollectionId;
             _client = cosmosClientInitializer.CreateCosmosClient(cosmosDataStoreConfiguration);
-
-            _initializationOperation = new RetryableInitializationOperation(
-                () => cosmosClientInitializer.InitializeDataStoreAsync(_client, cosmosDataStoreConfiguration, collectionInitializers));
-
             _container = new Lazy<Container>(() => cosmosClientInitializer.CreateFhirContainer(
                 _client,
                 cosmosDataStoreConfiguration.DatabaseId,
                 collectionId));
+            _initializationOperation = new RetryableInitializationOperation(async () =>
+            {
+                await collectionSetup.CreateDatabaseAsync(_client, cosmosDataStoreConfiguration, retryPolicyFactory.RetryPolicy, CancellationToken.None); // We need valid cancellation token
```

Comment:
I was thinking about this part of the code. 
We should encapsulate the call to these four methods in a single method with a try/catch block, and log every time it fails, because we need to know when the retries are happening.

Reply:
What I'm thinking is: we should have the same as in FhirCosmosClientInitliazer.InitializeDataStoreAsync: a try/catch that logs the exception and throws.

Reply:
done